presto-main/src/test/java/com/facebook/presto/sql/planner/TestLogicalPlanner.java

```diff
@@ -131,6 +138,155 @@ public void testAggregation()
+                                                                anyTree(tableScan("orders", ImmutableMap.of("orderstatus", "orderstatus", "orderkey", "orderkey"))),
+                                                                anyTree(tableScan("lineitem", ImmutableMap.of("linestatus", "linestatus"))))))))));
+
+        assertDistributedPlan("SELECT orderkey FROM (SELECT o.orderkey, row_number() OVER (PARTITION BY o.orderkey ORDER BY o.custkey) row_number FROM orders o JOIN lineitem l ON o.orderstatus = l.linestatus) WHERE row_number = 1",
+                anyTree(
+                        topNRowNumber(topNRowNumberMatcherBuilder -> topNRowNumberMatcherBuilder
+                                        .specification(
+                                                ImmutableList.of("orderkey"),
+                                                ImmutableList.of("custkey"),
+                                                ImmutableMap.of("custkey", ASC_NULLS_LAST))
+                                        .partial(false),
+                                exchange(LOCAL, GATHER,
+                                        exchange(REMOTE, REPARTITION,
+                                                topNRowNumber(topNRowNumberMatcherBuilder -> topNRowNumberMatcherBuilder
+                                                                .specification(
+                                                                        ImmutableList.of("orderkey"),
+                                                                        ImmutableList.of("custkey"),
+                                                                        ImmutableMap.of("custkey", ASC_NULLS_LAST))
+                                                                .partial(true),
+                                                        project(
+                                                                join(INNER, ImmutableList.of(equiJoinClause("orderstatus", "linestatus")),
+                                                                        anyTree(tableScan("orders", ImmutableMap.of("orderstatus", "orderstatus", "orderkey", "orderkey", "custkey", "custkey"))),
+                                                                        anyTree(tableScan("lineitem", ImmutableMap.of("linestatus", "linestatus")))))))))));
+    }
+
+    @Test
+    public void testWindowAfterAggregation()
+    {
+        // Window Partition key is a super set of group by key.
+        assertDistributedPlan("SELECT rank() OVER (PARTITION BY custkey) FROM orders GROUP BY custkey",
+                anyTree(
+                        window(windowMatcherBuilder -> windowMatcherBuilder
+                                        .specification(specification(ImmutableList.of("custkey"), ImmutableList.of(), ImmutableMap.of()))
+                                        .addFunction(functionCall("rank", Optional.empty(), ImmutableList.of())),
+                                anyNot(ExchangeNode.class,
```

Comment:
Why not to check that the input is an inmemory gather + aggregation? Same comment applies for the remaining test cases.